File: mindquantum/parameterresolver/parameterresolver.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Parameter resolver."""

from collections.abc import Iterable
from copy import deepcopy
import numpy as np
import sympy as sp


class ParameterResolver(dict):
    """
    A ParameterRsolver can set the parameter of parameterized quantum gate or
    parameterized quantum circuit.

    By specific which part of parameters needs to calculate gradient, the PQC
    operator can only calculate gradient of these parameters.

    Args:
        data (dict): initial parameter names and its values. Default: None.

    Examples:
        >>> from mindquantum import ParameterResolver
        >>> pr = ParameterResolver({'a': 0.3})
        >>> pr['b'] = 0.5
        >>> pr.no_grad_part('a')
        >>> pr *= 2
        >>> pr
        {'a': 0.6, 'b': 1.0}
        >>> pr.no_grad_parameters
        {'a'}
    """
    def __init__(self, data=None):
        if data is None:
            data = {}
        if not isinstance(data, (dict, ParameterResolver)):
            raise TypeError(
                "Require a dict or a ParameterResolver, but get {}!".format(
                    type(data)))
        for k, v in data.items():
            if not isinstance(k, str):
                raise TypeError(
                    "Parameter name should be a string, but get {}!".format(
                        type(k)))
            if not isinstance(v, _num_type):
                raise TypeError(
                    "Require a number, but get {}, which is {}!".format(
                        v, type(v)))
        super(ParameterResolver, self).__init__(data)
        self.no_grad_parameters = set()
        self.requires_grad_parameters = set(self.para_name)

    def __setitem__(self, keys, values):
        """
        Set parameter or as list of parameters of this parameter resolver.

        By default, the parameter you set requires gradient.

        Args:
            keys (Union[str, list[str]]): The name of parameters.
            values (Union[number, list[number]]): The value of parameters.

        Raises:
            TypeError: If the key that you set is not a string or a iterable of
                string.
        """
        if isinstance(keys, str):
            if not isinstance(values, _num_type):
                raise TypeError(
                    "Parameter value should be a number, but get {}, which is {}!"
                    .format(values, type(values)))
            super().__setitem__(keys, values)
            self.requires_grad_parameters.add(keys)
        elif isinstance(keys, Iterable):
            if not isinstance(values, Iterable):
                raise ValueError("Values should be iterable.")
            if len(values) != len(keys):
                raise ValueError("Size of keys and values do not match.")
            for i, k in enumerate(keys):
                self.__setitem__(k, values[i])
        else:
            raise TypeError(
                "Parameter name should be a string, but get {}!".format(
                    type(keys)))

    def __add__(self, pr):
        """
        Add a parameter resolver with other parameter.

        Returns:
            ParameterResolver, parameter resolver after adding.

        Args:
            pr (ParameterResolver): The parameter resolver need to add.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr1 = ParameterResolver({'a': 1})
            >>> pr2 = ParameterResolver({'a': 2, 'b': 3})
            >>> (pr1 + pr2).expression()
            3*a + 3*b
        """
        if not isinstance(pr, ParameterResolver):
            raise ValueError(
                'Require a parameter resolver, but get {}.'.format(type(pr)))
        res = self * 1
        pr = pr * 1
        for k, v in pr.items():
            if k in res:
                res[k] += v
                pr[k] = res[k]
        res.update(pr)
        return res

    def __sub__(self, pr):
        """
        Subtraction a parameter resolver with other parameter.

        Returns:
            :class:`mindquantum.parameterresolver.ParameterResolver`

        Args:
            pr (ParameterResolver): The parameter resolver need to subtract.

        Examples:
            >>> pr1 = ParameterResolver({'a': 1})
            >>> pr2 = ParameterResolver({'a': 2, 'b': 3})
            >>> (pr1 - pr2).expression()
            -a - 3*b
        """
        return self + (-1 * pr)

    def __neg__(self):
        """
        Get the negative version of this parameter resolver.

        Returns:
            ParameterResolver, the negative version.

        Examples:
            >>> pr1 = ParameterResolver({'a': 1})
            >>> (-pr1).expression()
            -a
        """
        return -1 * self

    def __imul__(self, num):
        """
        Parameter support inplace multiply.

        Returns:
            :class:`mindquantum.parameterresolver.ParameterResolver`

        Args:
            num (number): Multiply factor.

        Examples:
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr *= 2
            >>> pr
            {'a': 2, 'b': 4}
        """
        no_grad_parameters = deepcopy(self.no_grad_parameters)
        requires_grad_parameters = deepcopy(self.requires_grad_parameters)
        for k in self.keys():
            self[k] = self[k] * num
        self.no_grad_parameters = no_grad_parameters
        self.requires_grad_parameters = requires_grad_parameters
        return self

    def __mul__(self, num):
        """
        Multiply num with every value of parameter resolver.

        Returns:
            :class:`mindquantum.parameterresolver.ParameterResolver`

        Args:
            num (number): Multiply factor.

        Examples:
            >>> pr1 = ParameterResolver({'a': 1, 'b': 2})
            >>> pr2 = pr1 * 2
            >>> pr2
            {'a': 2, 'b': 4}
        """
        no_grad_parameters = deepcopy(self.no_grad_parameters)
        requires_grad_parameters = deepcopy(self.requires_grad_parameters)
        out = deepcopy(self)
        out *= num
        out.no_grad_parameters = no_grad_parameters
        out.requires_grad_parameters = requires_grad_parameters
        return out

    def __rmul__(self, num):
        """
        See :class:`mindquantum.parameterresolver.ParameterResolver.__mul__`.
        """
        return self.__mul__(num)

    def __eq__(self, other):
        _check_pr_type(other)
        no_grad_eq = self.no_grad_parameters == other.no_grad_parameters
        requires_grad_eq = self.requires_grad_parameters == other.requires_grad_parameters
        return super().__eq__(other) and no_grad_eq and requires_grad_eq

    @property
    def para_name(self):
        """
        Get the parameters name.

        Returns:
            list, a list of parameters name.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.para_name
            ['a', 'b']
        """
        return list(self.keys())

    @property
    def para_value(self):
        """
        Get the parameters value.

        Returns:
            list, a list of parameters value.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.para_value
            [1, 2]
        """
        return list(self.values())

    def requires_grad(self):
        """
        Set all parameters of this parameter resolver to require gradient
        calculation. Inplace operation.

        Returns:
            ParameterResolver, the parameter resolver itself.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.no_grad_part('a')
            >>> pr.requires_grad()
            >>> pr.requires_grad_parameters
            {'a', 'b'}
        """
        self.no_grad_parameters = set()
        self.requires_grad_parameters = set(self.para_name)
        return self

    def no_grad(self):
        """
        Set all parameters to not require gradient calculation. Inplace operation.

        Returns:
            ParameterResolver, the parameter resolver itself.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.no_grad()
            >>> pr.requires_grad_parameters
            set()
        """
        self.no_grad_parameters = set(self.para_name)
        self.requires_grad_parameters = set()
        return self

    def requires_grad_part(self, *names):
        """
        Set part of parameters that requires grad. Inplace operation.

        Args:
            names (tuple[str]): Parameters that requires grad.

        Returns:
            ParameterResolver, the parameter resolver itself.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.no_grad()
            >>> pr.requires_grad_part('a')
            >>> pr.requires_grad_parameters
            {'a'}
        """
        for name in names:
            if not isinstance(name, str):
                raise TypeError("name should be a string, but get {}!".format(
                    type(name)))
            if name not in self:
                raise KeyError(
                    "Parameter {} not in this parameter resolver!".format(
                        name))
            while name in self.no_grad_parameters:
                self.no_grad_parameters.remove(name)
            while name not in self.requires_grad_parameters:
                self.requires_grad_parameters.add(name)
        return self

    def no_grad_part(self, *names):
        """
        Set part of parameters that not requires grad.

        Args:
            names (tuple[str]): Parameters that not requires grad.

        Returns:
            ParameterResolver, the parameter resolver itself.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr = ParameterResolver({'a': 1, 'b': 2})
            >>> pr.no_grad_part('a')
            >>> pr.requires_grad_parameters
            {'b'}
        """
        for name in names:
            if not isinstance(name, str):
                raise TypeError("name should be a string, but get {}!".format(
                    type(name)))
            if name not in self:
                raise KeyError(
                    "Parameter {} not in this parameter resolver!".format(
                        name))
            while name not in self.no_grad_parameters:
                self.no_grad_parameters.add(name)
            while name in self.requires_grad_parameters:
                self.requires_grad_parameters.remove(name)
        return self

    def update(self, others):
        """
        Update this parameter resolver with other parameter resolver.

        Args:
            others (ParameterResolver): other parameter resolver.

        Raises:
            ValueError: If some parameters require grad and not require grad in
                other parameter resolver and vise versa.

        Examples:
            >>> from mindquantum import ParameterResolver
            >>> pr1 = ParameterResolver({'a': 1})
            >>> pr2 = ParameterResolver({'b': 2})
            >>> pr2.no_grad()
            >>> pr1.update(pr2)
            >>> pr1
            {'a': 1, 'b': 2}
            >>> pr1.no_grad_parameters
            {'b'}
        """
        _check_pr_type(others)
        super().update(others)
        conflict = (self.no_grad_parameters & others.requires_grad_parameters
                    ) | (others.no_grad_parameters
                         & self.requires_grad_parameters)
        if conflict:
            raise ValueError(
                "Parameter conflict, {} require grad in some parameter \
resolver and not require grad in other parameter resolver ".format(conflict))
        self.no_grad_parameters.update(others.no_grad_parameters)
        self.requires_grad_parameters.update(others.requires_grad_parameters)

    def mindspore_data(self):
        """
        Generate data for PQC operator.

        Returns:
            Dict.
        """
        m_data = {
            'gate_params_names': [],
            'gate_coeff': [],
            'gate_requires_grad': []
        }
        for k, v in self.items():
            m_data['gate_params_names'].append(k)
            m_data['gate_coeff'].append(float(v))
            m_data['gate_requires_grad'].append(
                k in self.requires_grad_parameters)
        return m_data

    def expression(self):
        """
        Get the expression of this parameter resolver.

        Returns:
            sympy.Expr, the symbol expression of this parameter resolver.

        Examples:
            >>> from mindquantum.parameterresolver import ParameterResolver as PR
            >>> pr = PR({'a' : 2, 'b' : 0.3})
            >>> pr.expression()
            2*a + 0.3*b
        """
        res = 0
        for k, v in self.items():
            res += sp.Symbol(k) * v
        return res

    def conjugate(self):
        """
        Get the conjugate of the parameter resolver.

        Returns:
            ParameterResolver, the conjugate version of this parameter resolver.

        Examples:
            >>> from mindquantum.parameterresolver import ParameterResolver as PR
            >>> pr = PR({'a' : 1, 'b': 1j})
            >>> pr.conjugate().expression()
            a - 1.0*I*b
        """
        out = 1 * self
        for k, v in out.items():
            out[k] = np.conj(v)
        return out

    def combination(self, pr):
        """
        Apply linear combination between this parameter resolver with input pr.

        Args:
            pr (Union[dict, ParameterResolver]): The parameter resolver you
                want to do linear combination.

        Returns:
            numbers.Number, the combination result.

        Examples:
        >>> from mindquantum import ParameterResolver
        >>> pr1 = ParameterResolver({'a': 1, 'b': 2})
        >>> pr2 = ParameterResolver({'a': 2, 'b': 3})
        >>> pr1.combination(pr2)
        """
        if not isinstance(pr, (ParameterResolver, dict)):
            raise ValueError(
                'Require a parameter resolver or a dict, but get {}.'.format(
                    type(pr)))
        res = 0
        for k, v in self.items():
            if k not in pr:
                raise KeyError('{} not in input parameter resolver'.format(k))
            res += v * pr[k]
        return res

    @property
    def real(self):
        """
        Get the real part of this parameter resolver

        Returns:
            ParameterResolver, the real part of this parameter resolver.

        Examples:
            >>> from mindquantum.parameterresolver import ParameterResolver as PR
            >>> pr = PR({'a': 1.2 + 1.3j})
            >>> pr.real()
            {'a': 1.2}
        """
        out = 1 * self
        for k, v in self.items():
            out[k] = np.real(v)
        return out

    @property
    def imag(self):
        """
        Get the real part of this parameter resolver

        Returns:
            ParameterResolver, the image part of this parameter resolver.

        Examples:
            >>> from mindquantum.parameterresolver import ParameterResolver as PR
            >>> pr = PR({'a': 1.2 + 1.3j})
            >>> pr.imag()
            {'a': 1.3}
        """
        out = 1 * self
        for k, v in self.items():
            out[k] = np.imag(v)
        return out


def _check_pr_type(pr):
    if not isinstance(pr, ParameterResolver):
        raise TypeError("Require a ParameterResolver, but get {}".format(
            type(pr)))


_num_type = (int, float, complex, np.int32, np.int64, np.float32, np.float64)

File: ccsrc/lib/math/pr/parameter_resolver.cpp
//   Copyright 2023 <Huawei Technologies Co., Ltd>
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

#include "math/pr/parameter_resolver.hpp"

#include <iostream>
#include <stdexcept>

#include "math/tensor/ops/advance_math.hpp"
#include "math/tensor/ops/memory_operator.hpp"
#include "math/tensor/tensor.hpp"
#include "math/tensor/traits.hpp"

namespace parameter {
ParameterResolver::ParameterResolver(const tn::Tensor& const_value) : const_value(const_value) {
}

ParameterResolver::ParameterResolver(const std::string& key, const tn::Tensor& const_value, tn::TDtype dtype) {
    this->const_value = const_value.astype(dtype);
    this->data_[key] = tn::ops::ones(1, dtype);
}

ParameterResolver::ParameterResolver(const std::map<std::string, tn::Tensor>& data, const tn::Tensor& const_value,
                                     tn::TDtype dtype) {
    this->const_value = const_value.astype(dtype);
    for (auto& [k, v] : data) {
        this->data_[k] = v.astype(dtype);
    }
}

tn::TDtype ParameterResolver::GetDtype() const {
    return this->const_value.dtype;
}

size_t ParameterResolver::Size() const {
    return this->data_.size();
}

void ParameterResolver::CastTo(tn::TDtype dtype) {
    if (dtype == this->GetDtype()) {
        return;
    }
    this->const_value = tn::ops::cast_to(this->const_value, dtype);
    for (auto& [k, v] : this->data_) {
        v = tn::ops::cast_to(v, dtype);
    }
}
void ParameterResolver::SetConstValue(const tn::Tensor& a) {
    this->const_value = a.astype(this->const_value.dtype);
}

tn::Tensor ParameterResolver::GetConstValue() const {
    return this->const_value;
}

std::string ParameterResolver::ToString() const {
    std::string out = "ParameterResolver(dtype: " + tensor::to_string(this->const_value.dtype) + ",";
    if (this->data_.size() == 0) {
        out += " const: " + tn::ops::to_string(this->const_value, true);
        out += ")";
        return out;
    }
    out += "\n";
    if (this->data_.size() == 0) {
        out += "  data: [],\n";
    } else {
        out += "  data: [\n";
        int i = 0;
        for (auto& [k, v] : this->data_) {
            out += "         " + k + ": " + tn::ops::to_string(v, true);
            i += 1;
            if (i != this->data_.size()) {
                out += ",";
            }
            out += "\n";
        }
        out += "  ],\n";
    }
    out += "  const: " + tn::ops::to_string(this->const_value, true);
    if (this->no_grad_parameters_.size() != 0) {
        out += ",\n  no grad parameters: {";
        int i = 0;
        for (auto& v : this->no_grad_parameters_) {
            out += v;
            i += 1;
            if (i != this->data_.size()) {
                out += ", ";
            }
        }
        out += "}";
    }
    if (this->encoder_parameters_.size() != 0) {
        out += ",\n  encoder parameters: {";
        int i = 0;
        for (auto& v : this->encoder_parameters_) {
            out += v;
            i += 1;
            if (i != this->data_.size()) {
                out += ", ";
            }
        }
        out += "}";
    }
    out += "\n)";
    return out;
}

bool ParameterResolver::Contains(const std::string& key) const {
    auto res = this->data_.find(key) != this->data_.end();
    return res;
}

bool ParameterResolver::NoGradContains(const std::string& key) const {
    return this->no_grad_parameters_.find(key) != this->no_grad_parameters_.end();
}

bool ParameterResolver::EncoderContains(const std::string& key) const {
    return this->encoder_parameters_.find(key) != this->encoder_parameters_.end();
}

std::set<std::string> ParameterResolver::GetAllParameters() const {
    std::set<std::string> out{};
    for (auto& [k, v] : this->data_) {
        out.insert(k);
    }
    return out;
}

std::set<std::string> ParameterResolver::GetRequiresGradParameters() const {
    return this->GetAllParameters() - this->no_grad_parameters_;
}

std::set<std::string> ParameterResolver::GetAnsatzParameters() const {
    return this->GetAllParameters() - this->encoder_parameters_;
}

bool ParameterResolver::IsConst() const {
    for (auto& [k, v] : this->data_) {
        if (!tn::ops::is_all_zero(v)) {
            return false;
        }
    }
    return true;
}

bool ParameterResolver::IsNotZero() const {
    if (!tn::ops::is_all_zero(this->const_value)) {
        return true;
    }
    for (auto& [k, v] : this->data_) {
        if (!tn::ops::is_all_zero(v)) {
            return true;
        }
    }
    return false;
}
std::vector<std::string> ParameterResolver::subs(const ParameterResolver& other) {
    std::vector<std::string> will_pop;
    auto origin_dtype = this->GetDtype();
    for (auto& [k, v] : this->data_) {
        if (other.Contains(k)) {
            this->const_value = this->const_value + v * other.data_.at(k);
            will_pop.push_back(k);
        }
    }
    if (will_pop.size() != 0) {
        for (auto& k : will_pop) {
            this->Pop(k);
        }
        auto new_dtype = this->GetDtype();
        if (origin_dtype != new_dtype) {
            for (auto& [k, v] : this->data_) {
                v = v.astype(new_dtype);
            }
        }
    }
    return will_pop;
}

tn::Tensor ParameterResolver::GetItem(const std::string& key) const {
    if (!this->Contains(key)) {
        throw std::runtime_error("parameter " + key + " not in this parameter resolver.");
    }
    return this->data_.at(key);
}

// -----------------------------------------------------------------------------

ParameterResolver& ParameterResolver::operator+=(const ParameterResolver& rhs) {
    if (((this->encoder_parameters_ & rhs.GetAnsatzParameters()).size() != 0)
        || ((this->GetAnsatzParameters() & rhs.encoder_parameters_).size() != 0)) {
        throw std::runtime_error("encoder or ansatz property of parameter conflict.");
    }
    if (((this->no_grad_parameters_ & rhs.GetRequiresGradParameters()).size() != 0)
        || ((this->GetRequiresGradParameters() & rhs.no_grad_parameters_).size() != 0)) {
        throw std::runtime_error("gradient property of parameter conflict.");
    }
    for (auto& [key, value] : rhs.data_) {
        if (this->Contains(key)) {
            this->data_[key] += value;
        } else {
            this->SetItem(key, value);
            if (rhs.EncoderContains(key)) {
                this->encoder_parameters_.insert(key);
            }
            if (rhs.NoGradContains(key)) {
                this->no_grad_parameters_.insert(key);
            }
        }
    }
    this->const_value += rhs.const_value;
    return *this;
}

ParameterResolver& ParameterResolver::operator-=(const ParameterResolver& rhs) {
    if (((this->encoder_parameters_ & rhs.GetAnsatzParameters()).size() != 0)
        || ((this->GetAnsatzParameters() & rhs.encoder_parameters_).size() != 0)) {
        throw std::runtime_error("encoder or ansatz property of parameter conflict.");
    }
    if (((this->no_grad_parameters_ & rhs.GetRequiresGradParameters()).size() != 0)
        || ((this->GetRequiresGradParameters() & rhs.no_grad_parameters_).size() != 0)) {
        throw std::runtime_error("gradient property of parameter conflict.");
    }
    for (auto& [key, value] : rhs.data_) {
        if (this->Contains(key)) {
            this->data_[key] -= value;
        } else {
            this->SetItem(key, static_cast<float>(0.0) - value);
            if (rhs.EncoderContains(key)) {
                this->encoder_parameters_.insert(key);
            }
            if (rhs.NoGradContains(key)) {
                this->no_grad_parameters_.insert(key);
            }
        }
    }
    this->const_value -= rhs.const_value;
    return *this;
}

ParameterResolver& ParameterResolver::operator*=(const ParameterResolver& rhs) {
    if (this->IsConst()) {
        for (auto& [k, v] : rhs.data_) {
            this->data_[k] = this->const_value * v;
            if (!this->Contains(k)) {
                if (rhs.EncoderContains(k)) {
                    this->encoder_parameters_.insert(k);
                }
                if (rhs.NoGradContains(k)) {
                    this->no_grad_parameters_.insert(k);
                }
            }
        }
    } else if (rhs.IsConst()) {
        for (auto& [k, v] : this->data_) {
            this->data_[k] *= rhs.const_value;
        }
    } else {
        throw std::runtime_error("Parameter resolver only support first order variable.");
    }

    this->const_value *= rhs.const_value;
    return *this;
}

ParameterResolver& ParameterResolver::operator/=(const ParameterResolver& rhs) {
    if (!rhs.IsConst()) {
        throw std::runtime_error("Cannot div a non constant ParameterResolver.");
    }
    for (auto& [k, v] : this->data_) {
        this->data_[k] /= rhs.const_value;
    }
    this->const_value /= rhs.const_value;
    return *this;
}

bool ParameterResolver::operator==(const ParameterResolver& value) {
    if (this->data_.size() != value.data_.size()) {
        return false;
    }
    if (!tn::ops::all_equal_to(this->const_value, value.const_value)) {
        return false;
    }
    for (auto& [k, v] : this->data_) {
        if (!value.Contains(k)) {
            return false;
        }
        if (!tn::ops::all_equal_to(v, this->data_[k])) {
            return false;
        }
    }
    return true;
}

bool ParameterResolver::operator!=(const ParameterResolver& value) {
    return !(*this == value);
}

ParameterResolver ParameterResolver::operator-() const {
    auto out = *this;
    out.const_value = 0.0 - out.const_value;
    for (auto& [k, v] : out.data_) {
        v = 0.0 - v;
    }
    return out;
}

ParameterResolver operator+(const ParameterResolver& lhs, const ParameterResolver& rhs) {
    auto origin_type = lhs.const_value.dtype;
    auto out = lhs;
    auto new_const = lhs.const_value + rhs.const_value;
    auto new_type = new_const.dtype;
    if (new_type != origin_type) {
        out.CastTo(new_type);
    }
    out += rhs;
    return out;
}

ParameterResolver operator-(const ParameterResolver& lhs, const ParameterResolver& rhs) {
    auto origin_type = lhs.const_value.dtype;
    auto out = lhs;
    auto new_const = lhs.const_value + rhs.const_value;
    auto new_type = new_const.dtype;
    if (new_type != origin_type) {
        out.CastTo(new_type);
    }
    out -= rhs;
    return out;
}

ParameterResolver operator*(const ParameterResolver& lhs, const ParameterResolver& rhs) {
    auto origin_type = lhs.const_value.dtype;
    auto out = lhs;
    auto new_const = lhs.const_value + rhs.const_value;
    auto new_type = new_const.dtype;
    if (new_type != origin_type) {
        out.CastTo(new_type);
    }
    out *= rhs;
    return out;
}

ParameterResolver operator/(const ParameterResolver& lhs, const ParameterResolver& rhs) {
    auto origin_type = lhs.const_value.dtype;
    auto out = lhs;
    auto new_const = lhs.const_value + rhs.const_value;
    auto new_type = new_const.dtype;
    if (new_type != origin_type) {
        out.CastTo(new_type);
    }
    out /= rhs;
    return out;
}

// -----------------------------------------------------------------------------

std::vector<std::string> ParameterResolver::ParamsName() const {
    std::vector<std::string> out = {};
    for (auto& [k, v] : this->data_) {
        out.push_back(k);
    }
    return out;
}

std::vector<tn::Tensor> ParameterResolver::ParaValue() const {
    std::vector<tn::Tensor> out = {};
    for (auto& [k, v] : this->data_) {
        out.push_back(v);
    }
    return out;
}
auto ParameterResolver::ParaData() const -> data_t {
    return this->data_;
}

void ParameterResolver::RequiresGrad() {
    this->no_grad_parameters_ = {};
}

void ParameterResolver::NoGrad() {
    this->no_grad_parameters_ = {};
    for (auto& [k, v] : this->data_) {
        this->no_grad_parameters_.insert(k);
    }
}

void ParameterResolver::RequiresGradPart(const std::vector<std::string>& names) {
    for (auto& name : names) {
        if (this->NoGradContains(name)) {
            this->no_grad_parameters_.erase(name);
        }
    }
}

void ParameterResolver::NoGradPart(const std::vector<std::string>& names) {
    for (auto& name : names) {
        if (this->Contains(name)) {
            this->no_grad_parameters_.insert(name);
        }
    }
}

void ParameterResolver::AnsatzPart(const std::vector<std::string>& names) {
    for (auto& name : names) {
        if (this->EncoderContains(name)) {
            this->encoder_parameters_.erase(name);
        }
    }
}

void ParameterResolver::EncoderPart(const std::vector<std::string>& names) {
    for (auto& name : names) {
        if (this->Contains(name)) {
            this->encoder_parameters_.insert(name);
        }
    }
}

void ParameterResolver::AsEncoder() {
    for (auto& [k, v] : this->data_) {
        this->encoder_parameters_.insert(k);
    }
}

void ParameterResolver::AsAnsatz() {
    this->encoder_parameters_ = {};
}

void ParameterResolver::Update(const ParameterResolver& other) {
    if (((this->encoder_parameters_ & other.GetAnsatzParameters()).size() != 0)
        | ((this->GetAnsatzParameters() & other.encoder_parameters_).size() != 0)) {
        throw std::runtime_error("encoder or ansatz property of parameter conflict.");
    }
    if (((this->no_grad_parameters_ & other.GetRequiresGradParameters()).size() != 0)
        | ((this->GetRequiresGradParameters() & other.no_grad_parameters_).size() != 0)) {
        throw std::runtime_error("gradient property of parameter conflict.");
    }

    for (auto& [key, value] : other.data_) {
        if (this->Contains(key)) {
            this->SetItem(key, value);
        } else {
            this->SetItem(key, value);
            if (other.EncoderContains(key)) {
                this->encoder_parameters_.insert(key);
            }
            if (other.NoGradContains(key)) {
                this->no_grad_parameters_.insert(key);
            }
        }
    }
    this->const_value = other.const_value;
}

ParameterResolver ParameterResolver::Conjugate() const {
    auto out = *this;
    out.const_value = out.const_value.conj();
    for (auto& [k, v] : out.data_) {
        v = v.conj();
    }
    return out;
}

ParameterResolver ParameterResolver::Combination(const ParameterResolver& pr) const {
    auto c = this->const_value;
    for (auto& [k, v] : this->data_) {
        c += v * pr.GetItem(k);
    }
    auto out = ParameterResolver();
    out.const_value = c;
    return out;
}

ParameterResolver ParameterResolver::Real() const {
    auto out = *this;
    out.const_value = out.const_value.real();
    for (auto& [k, v] : out.data_) {
        v = v.real();
    }
    return out;
}

void ParameterResolver::KeepReal() {
    this->const_value = this->const_value.real();
    for (auto& [k, v] : this->data_) {
        v = v.real();
    }
}

void ParameterResolver::KeepImag() {
    this->const_value = this->const_value.imag();
    for (auto& [k, v] : this->data_) {
        v = v.imag();
    }
}

ParameterResolver ParameterResolver::Imag() const {
    auto out = *this;
    out.const_value = out.const_value.imag();
    for (auto& [k, v] : out.data_) {
        v = v.imag();
    }
    return out;
}

tn::Tensor ParameterResolver::Pop(const std::string& key) {
    auto out = this->GetItem(key);
    this->data_.erase(key);
    if (this->EncoderContains(key)) {
        this->encoder_parameters_.erase(key);
    }
    if (this->NoGradContains(key)) {
        this->no_grad_parameters_.erase(key);
    }
    return out;
}

bool ParameterResolver::IsHermitian() const {
    return !(*this - this->Conjugate()).IsNotZero();
}

bool ParameterResolver::IsAntiHermitian() const {
    return !(*this + this->Conjugate()).IsNotZero();
}

bool ParameterResolver::HasRequireGradParams() {
    return this->data_.size() > this->no_grad_parameters_.size();
}
std::map<std::string, size_t> GetRequiresGradParameters(const std::vector<ParameterResolver>& prs) {
    std::map<std::string, size_t> title = {};
    size_t idx = 0;
    for (auto& pr : prs) {
        for (auto& name : pr.GetRequiresGradParameters()) {
            if (!title.count(name)) {
                title[name] = idx;
                ++idx;
            }
        }
    }
    return title;
}

std::pair<std::map<std::string, size_t>, tensor::Matrix> Jacobi(const std::vector<ParameterResolver>& prs) {
    auto title = GetRequiresGradParameters(prs);
    auto upper_t = tensor::TDtype::Float32;
    for (auto& pr : prs) {
        upper_t = tensor::upper_type_v(upper_t, pr.GetDtype());
    }
    auto jacobi = tensor::ops::zeros(prs.size() * title.size());
    for (size_t i = 0; i < prs.size(); i++) {
        for (auto& name : prs[i].GetRequiresGradParameters()) {
            tensor::ops::cpu::set(&jacobi, prs[i].data_.at(name), i * title.size() + title.at(name));
        }
    }
    return {title, tensor::Matrix(std::move(jacobi), prs.size(), title.size())};
}
}  // namespace parameter
std::ostream& operator<<(std::ostream& os, const parameter::ParameterResolver& pr) {
    os << pr.ToString();
    return os;
}
